src/algebra/algebra/basic.lean

```diff
@@ -1404,20 +1404,28 @@ end linear_map
 end is_scalar_tower
 
 section restrict_scalars
-/- In this section, we describe restriction of scalars: if `S` is an algebra over `R`, then
-`S`-modules are also `R`-modules. -/
 
 section type_synonym
 variables (R A M : Type*)
 
-/--
-Warning: use this type synonym judiciously!
-The preferred way of working with an `A`-module `M` as `R`-module (where `A` is an `R`-algebra),
-is by `[module R M] [module A M] [is_scalar_tower R A M]`.
+/-- A type synonym representing the functor from the category of modules over `A`, regarded as a
+(semi)ring, to the category of modules over `A`, regarded as an `R`-algebra.
 
-When `M` is a module over a ring `A`, and `A` is an algebra over `R`, then `M` inherits a
-module structure over `R`, provided as a type synonym `module.restrict_scalars R A M := M`.
--/
+Warning: use this type synonym judiciously! Often what one wants instead of `restrict_scalars` is
+actually the following setup:
+`[comm_semiring R] [semiring A] [algebra R A] [module R M] [module A M] [is_scalar_tower R A M]` (*)
+which is just the way of saying "let `M` be a module over the `R`-algebra `A`".
+
+Mathematically, when `A` is an `R`-algebra, there are two different functors:
+ 1. The functor from the category of modules over the (semi)ring `A` to the category of modules over
+    the `R`-algebra `A`. This is what is defined here, and is called `restrict_scalars` (though
+    perhaps a better name would be `enrich_scalars`).
+ 2. The functor from the category of modules over the `R`-algebra `A` to the category of modules
+    over the (semi)ring `R`. This is the true restriction of scalars but when one uses the setup (*)
+    above, it is invisible functor available for free through typeclass inference.
+
+A standard example of when one might want to invoke 1 would be if one has a vector space over a
+field of characteristic zero and wishes to make use of the `ℚ`-algebra structure. -/
```

Comment:
This doesn't seem like a compelling justification; doesn't typeclass inference find the `ℚ`-algebra structure automatically?

Reply:
When it attempts the failed inference, it has no idea about `K` so it cannot magic is up.

Reply:
Ah, I guess the `is_scalar_tower` solution here would be "just pass `module ℚ V` as a typeclass argument, which lowers the chance of diamonds.